src/app/services/auth.service.ts

```diff
@@ -0,0 +1,72 @@
+import {Injectable} from '@angular/core';
+import {Router} from '@angular/router';
+import {auth} from 'firebase/app';
+import {Observable} from 'rxjs';
+import {User} from '../models/user.model';
+import * as firebase from 'firebase';
+import {NotificationService} from './notification.service';
+import {of} from 'rxjs/internal/observable/of';
+import {environment} from '../../environments/environment';
+import {AngularFireAuth} from '@angular/fire/auth';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AuthService {
+  user$: Observable<User | null | undefined> = of(undefined);
+
+  constructor(
+    private router: Router,
+    private notificationService: NotificationService,
+    public angularFireAuth: AngularFireAuth
+  ) {
+    if (environment.requireAuth) {
+        angularFireAuth.onAuthStateChanged(firebaseUser => {
+          if (firebaseUser != null) {
+            const user: User = {uid: firebaseUser.uid, displayName: firebaseUser.displayName};
+            this.user$ = of(user);
```

Comment:
I'm a bit surprised that we'd create a brand-new `Observable` every time the user changes. I thought the point of an `Observable` was that it provides a stream of values, to which other components can then subscribe?

Presumably in this case, the intent is to enable other components to subscribe for who the current user is?

In that case I'd have thought we'd make `user$` a `readonly` member, and then initialize it in the constructor to something like a `BehaviorSubject`:

https://rxjs.dev/guide/subject#behaviorsubject

Then, you call `next()` on that subject with the new value inside your `onAuthStateChanged` handler?

This way, `user$` is a constant `Observable` which emits a new value every time the user changes.